### DISCUSSION TITLE
Bug in content negotiation when specifying a default format and including it in a call to respond_to() helper

### DISCUSSION THREAD

**Discussion by vmmello** [mojolicious/mojo#2187] [Q&A]
* Mojolicious version: 9.37
* Perl version: 5.38.2
* Operating system: Alpine Linux 3.19

I'm trying to change the default response type of an endpoint to `text/plain`, so that when a user runs `curl /` and the app receives the header `Accept: */*` it will reply with a `txt`, not html.

A simplified sample code of how I'm trying it:

```perl
get '/'  => { format => "txt" } => sub($c) {
              $c->respond_to(
                html => { },
                txt   => { },
                json  => { },
                any  => { format => "txt" },
              );
        }
        => "index";

__DATA__

@@ index.html.ep
<html></html>

@@ index.txt.ep
text

@@ index.json.ep
{"json":""}

```

The above looks right to me. But it completely breaks content negotiation, replying all requests with the default format. For example, when running:

`$ ./app.pl get -H 'Accept: text/html' /`
```
text
```

`$ ./app.pl get -H 'Accept: application/json' /`
```
text
```

But if I remove the `txt => {}` line (line 4) from `respond_to()`, letting `text/plain` be catched by the `any` format, it works as expected (i.e. content negotiation through `Accept` header works, and it correctly fallsback unknowns to `txt` format). If I omit the `any` format it replies with `204 No Content` for `txt` and other formats (other than the ones listed in `respond_to()`).

So I believe this is a subtle bug in content negotiation, the fact that when you explicitly specify a default format and include it in a `respond_to()` method call, it breaks content negotiation through the `Accept` header.

This same problem happens when the default format is explicitly specified as `html`, e.g.: 
`get '/' => { format => "html" } => sub { ... };`
so it's not the case of changing the response format to something other than `html`.

**Comment by kraih** [comment]
I don't quite see the bug i'm afraid, the logic seems to follow the documentation in the precedence order it is described. https://docs.mojolicious.org/Mojolicious/Guides/Rendering#Content-negotiation

**Comment by vmmello** [comment]
I insist on my initial point, though I'll break it down in a explicit example:

This works as expected:
```perl
get '/' => { format => "txt" } => sub($c) {
  $c->respond_to(
    html => {},
    json => {},
    any => { format => "txt"}
  );
} => "index";
```

This one below completely breaks content negotiation (replies everything with the default format and ignores the `Accept` header):

```perl
get '/' => { format => "txt" } => sub($c) {
  $c->respond_to(
  txt => {},
  html => {},
  json => {},
  any => { format => "txt" },
);
} => "index";
```

The simple fact that the default format is included inside the `respond_to()` call (like in this second example) it completely breaks content negotiation.

So I don't think this subtle change is the expected behavior.
